### PR TITLE
feat: preserve_host feature for oauth2_introspect

### PR DESCRIPTION
**Pull request by alnr**
- fixed lots of lint issues
- bump to Go 1.21
- fixed context passing, cancelation, and tracing in a couple of places
- added the `preserve_host` option which already exists in the `cookie_session` and `bearer_token` authenticators to the `oauth2_introspection` authenticator.
- added the `prefix` option to the `oauth2_introspection` and `bearer_token` authenticators. This allows you to filter out tokens based on a string prefix.

### CODE REVIEW

File: credentials/fetcher_default.go [hunk 1]
```diff
@@ -112,12 +120,12 @@ func (s *FetcherDefault) fetchParallel(ctx context.Context, locations []url.URL)
...
+	go s.resolveAll(ctx, done, errs, locations)
 
 	select {
-	case <-ctx.Done():
-		s.l.WithError(ctx.Err()).Errorf("Ignoring JSON Web Keys from at least one URI because the request timed out waiting for a response.")
-		return ctx.Err()
+	case <-time.After(s.cancelAfter):
+		s.l.Errorf("Ignoring JSON Web Keys from at least one URI because the request timed out waiting for a response.")
+		return context.DeadlineExceeded
 	case <-done:
```
Comment (aeneasr) [inline credentials/fetcher_default.go:129]: ctx.Done is now missing, no? The parent of `fetchParallel` might still cancel the context.

File: pipeline/authn/authenticator_jwt.go [hunk 2]
```diff
@@ -71,7 +74,11 @@ func (a *AuthenticatorJWT) Config(config json.RawMessage) (*AuthenticatorOAuth2J
 	return &c, nil
 }
 
-func (a *AuthenticatorJWT) Authenticate(r *http.Request, session *AuthenticationSession, config json.RawMessage, _ pipeline.Rule) error {
+func (a *AuthenticatorJWT) Authenticate(r *http.Request, session *AuthenticationSession, config json.RawMessage, _ pipeline.Rule) (err error) {
+	ctx, span := a.r.Tracer().Start(r.Context(), "authn.jwt")
```
Comment (aeneasr) [inline pipeline/authn/authenticator_jwt.go:78]: Please keep the semantics we use everywhere else:

```suggestion
	ctx, span := a.r.Tracer().Start(r.Context(), "pipeline.authn.AuthenticatorJWT.Authenticate")
```

File: pipeline/authn/authenticator_oauth2_introspection.go [hunk 3]
```diff
@@ -169,14 +175,19 @@ func (a *AuthenticatorOAuth2Introspection) tokenToCache(config *AuthenticatorOAu
 	}
 }
 
-func (a *AuthenticatorOAuth2Introspection) Authenticate(r *http.Request, session *AuthenticationSession, config json.RawMessage, _ pipeline.Rule) error {
+func (a *AuthenticatorOAuth2Introspection) Authenticate(r *http.Request, session *AuthenticationSession, config json.RawMessage, _ pipeline.Rule) (err error) {
+	tp := trace.SpanFromContext(r.Context()).TracerProvider()
+	ctx, span := tp.Tracer("oauthkeeper/pipeline/authn").Start(r.Context(), "authn.oauth2_introspection")
```
Comment (aeneasr) [inline pipeline/authn/authenticator_oauth2_introspection.go:180]:
```suggestion
	ctx, span := tp.Tracer("oauthkeeper/pipeline/authn").Start(r.Context(), "pipeline.authn.AuthenticatorOAuth2Introspection.Authenticate")
```

File: pipeline/authn/authenticator_oauth2_introspection.go [hunk 4]
```diff
@@ -204,7 +215,11 @@ func (a *AuthenticatorOAuth2Introspection) Authenticate(r *http.Request, session
 		// set/override the content-type header
 		introspectReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
 
-		resp, err := client.Do(introspectReq.WithContext(r.Context()))
+		if cf.PreserveHost {
```
Comment (aeneasr) [inline pipeline/authn/authenticator_oauth2_introspection.go:218]: Test please

**Review by aeneasr — APPROVED** [review]

**Comment by aeneasr** [comment]
Looks like the changes are causing issues with the cache!

File: credentials/fetcher_default.go (hunk 1 above)
Comment (alnr) [reply credentials/fetcher_default.go:131]: fixed

File: pipeline/authn/authenticator_jwt.go (hunk 2 above)
Comment (alnr) [reply pipeline/authn/authenticator_jwt.go]: unified everything in oathkeeper

**Comment by alnr** [comment]
> Looks like the changes are causing issues with the cache!

I've seen this flake before. Must be something internal to ristretto. Couldn't reproduce it locally, even with `go test -count=100` 🤷

File: spec/config.schema.json [hunk 5]
```diff
@@ -483,6 +483,11 @@
             }
           ]
         },
+        "prefix": {
+          "title": "Token Prefix",
+          "type": "string",
+          "description": "The token is considered only if starts with this string."
```
Comment (Benehiko) [inline spec/config.schema.json:489]:
```suggestion
          "description": "The token is considered only if it starts with this string."
```

File: spec/config.schema.json [hunk 6]
```diff
@@ -822,6 +827,17 @@
             "type": "string"
           }
         },
+        "prefix": {
+          "title": "Token Prefix",
+          "type": "string",
+          "description": "The token is considered only if starts with this string."
```
Comment (Benehiko) [inline spec/config.schema.json:833]:
```suggestion
          "description": "The token is considered only if it starts with this string."
```

**Review by Benehiko — APPROVED** [review]

**Review by aeneasr — APPROVED** [review]

**Pull request merged**